ghcide/src/Development/IDE/Types/Diagnostics.hs

```diff
@@ -44,26 +56,53 @@ type IdeResult v = ([FileDiagnostic], Maybe v)
 -- | an IdeResult with a fingerprint
 type IdeResultNoDiagnosticsEarlyCutoff  v = (Maybe ByteString, Maybe v)
 
-ideErrorText :: NormalizedFilePath -> T.Text -> FileDiagnostic
-ideErrorText = ideErrorWithSource (Just "compiler") (Just DiagnosticSeverity_Error)
+ideErrorText :: Maybe (MsgEnvelope GhcMessage) -> NormalizedFilePath -> T.Text -> FileDiagnostic
+ideErrorText origMsg fdFilePath msg =
+  ideErrorWithSource (Just "compiler") (Just DiagnosticSeverity_Error) fdFilePath msg origMsg
+
+ideErrorFromLspDiag
+  :: LSP.Diagnostic
+  -> NormalizedFilePath
+  -> Maybe (MsgEnvelope GhcMessage)
+  -> FileDiagnostic
+ideErrorFromLspDiag lspDiag fdFilePath origMsg =
+  let fdShouldShowDiagnostic = ShowDiag
+      fdStructuredMessage =
+        maybe NoStructuredMessage SomeStructuredMessage origMsg
+      fdLspDiagnostic = lspDiag
+        { _code = fmap ghcCodeToLspCode . diagnosticCode . errMsgDiagnostic =<< origMsg
```

Comment:
Hmm, we set it here but not below? confusing. Maybe these should share code?

Reply:
The diagnostic generating functions are all a bit of a pile (in the Enrico Fermi sense of the word) right now - let's discuss refactoring that in https://github.com/haskell/haskell-language-server/pull/4311#discussion_r1636130917